src/common/wayfinder/verification/data-root-verifier.ts

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (C) 2022-2024 Permanent Data Solutions, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  Chunk,
+  MAX_CHUNK_SIZE,
+  MIN_CHUNK_SIZE,
+  buildLayers,
+  generateLeaves,
+} from 'arweave/node/lib/merkle.js';
+import { Readable } from 'node:stream';
+
+import { DataRootProvider, DataVerifier } from '../../../types/wayfinder.js';
+import {
+  hashBufferToB64Url,
+  hashReadableStreamToB64Url,
+  hashReadableToB64Url,
+} from '../../../utils/hash.js';
+
+export async function convertBufferToDataRoot({
```

Comment:
Calculating data roots this way is fine, but it may not reproduce the original data root. We could also implement another data root verifier that pulls all the chunks, checks that they reproduce the expected data root, and then compares the concatenated chunk data to the original data retrieved. That would take a while, bit it should be able to verify any L1 data where we can find the chunks.